### PR TITLE
Unit tests for the CookieLooselyScopedScannerUnitTest class (issue #2800)

### PR DESCRIPTION
**Pull request by vahidR**
The commit includes the following:
1. Unit tests for the CookieLooselyScopedScannerUnitTest class
2. Two necessary refactorings to make the filtering of cookie domains more robust
3. Adding IntelliJ IDEA specific files to gitignore

Fixes zaproxy/zaproxy#2800

### CODE REVIEW

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java [hunk 1]
```diff
@@ -104,8 +104,8 @@ private boolean isLooselyScopedCookie(HttpCookie cookie, String host) {
 			return cookieDomain.equals(host);
 		}
 		
-		// otherwise, remove the '.' and compare the result with the host
-		if (cookieDomains.length != 2) {
+		// otherwise, remove the '.' (if cookie domain is not empty) and compare the result with the host
+		if (!cookieDomain.isEmpty() && cookieDomains.length != 2) {
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: If `cookieDomain.isEmpty()` wouldn't the code have returned at L91 and not gotten here?

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java [hunk 2]
```diff
@@ -85,10 +85,10 @@ private boolean isLooselyScopedCookie(HttpCookie cookie, String host) {
 		assert host != null;
 		
 		String cookieDomain = cookie.getDomain();
-		
+
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: Please revert this whitespace change. It could be committed separately if needed.
https://github.com/zaproxy/zaproxy/wiki/DevGuidelines

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java (hunk 1 above)
Comment (vahidR) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: Ah right... I refactored line 108 first and then came across an edge case in line 91 :)

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java (hunk 2 above)
Comment (vahidR) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: Should be one of those invisible lines on my IDE again... 
I will fix it.

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: No problem, happens to me sometime too :smile:

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,8 @@
 build/build/
 build/temp/
 build/zap-exts/
+build/results/
```
Comment (thc202) [inline .gitignore]: This entry is already added in the (upstream) beta branch (a rebase should take care of that).

File: test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+        // When
+        rule.scanHttpResponseReceive(msg, -1, this.createSource(msg));
+
+        // Then
+        assertThat(alertsRaised.size(), is(0));
+    }
+
+    @Test
+    public void shouldRaiseAlertIfHostDomainStartsWithDot() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java]: Why? Seems to be a bug in the scanner, statement at line 104 (`cookieDomain.equals(host);`) should be negated?

File: test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+        // When
+        rule.scanHttpResponseReceive(msg, -1, this.createSource(msg));
+
+        // Then
+        assertThat(alertsRaised.size(), is(1));
+    }
+
+    @Test
+    public void shouldRaiseAlertIfHostDomainAndCookieDomainAreTheSameAndDomainsAreNotSecondLevel() throws Exception {
```
Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java]: Same comment.

File: test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java (hunk 4 above)
Comment (vahidR) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java]: > Why? Seems to be a bug in the scanner

This is how it works right now. If it is not intended, it should be refactored

> statement at line 104 (cookieDomain.equals(host);) should be negated?

I was a bit skeptical to do that. 
There is another case which is not covered: IF cookie domain and host domain are different! Although web servers mostly neglect this (IMO), but it's good to cover that too..

File: test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java (hunk 5 above)
Comment (vahidR) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java]: This logic confused me too... I imagined this is an intended behaviour if we have more levels

File: test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java (hunk 4 above)
Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java]: Right, was questioning the behaviour of the scanner not the test. Probably better to fix that now.

Agreed, should detect and alert on that case too.

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java [hunk 6]
```diff
@@ -83,12 +83,12 @@ private boolean isLooselyScopedCookie(HttpCookie cookie, String host) {
...
 		assert host != null;
-		
+
 		String cookieDomain = cookie.getDomain();
-		
+
 		// if Domain attribute hasn't been specified, the cookie
 		// is scoped with the response host
-		if (cookieDomain == null) {
+		if (cookieDomain == null || cookieDomain.equalsIgnoreCase("null") || cookieDomain.isEmpty()) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: When does the "null" case happen?

**Comment by thc202** [comment]
Worth having a test(s) to assert that the domain comparison is case insensitive.

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java (hunk 6 above)
Comment (vahidR) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: I must have removed the test for that edge case somehow, but one way to reproduce it is `Set-Cookie: a=b;domain=null` or having a malformed domain `Set-Cookie: a=b;domain`.
I will write down that edge case, if I remember it..

File: test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java (hunk 4 above)
Comment (vahidR) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java]: > Right, was questioning the behaviour of the scanner not the test. Probably better to fix that now.

I'll refactor the code..

Comment (vahidR) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java]: @thc202 shouldn't the code `cookieDomain.equals(host);` be reversed ? 
IF `cookie domain` and `host domain` are the same (in this  case), then it is not Loosed? Negating the logic makes more sense!

Comment (thc202) [inline test/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScannerUnitTest.java]: Yeah, that's my understanding too.

File: src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java (hunk 6 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: Well, having the domain set to "null" should not be treated differently than any other value. For the latter case, the domain is only "null" if the request header does not have a target host (which should have).

Comment (vahidR) [inline src/org/zaproxy/zap/extension/pscanrulesBeta/CookieLooselyScopedScanner.java]: Fair enough ... I will remove the `"null"` check then

**Comment by vahidR** [comment]
> Worth having a test(s) to assert that the domain comparison is case insensitive.

I added the unit test for that..

**Comment by kingthorin** [comment]
Looks good to me

**Comment by kingthorin** [comment]
TODO:

- [ ] Add manifest update
- [ ] Extract a method for the creation of a "standard" HttpMessage
- [ ] Revert whitespace changes

**Comment by thc202** [comment]
@kingthorin, @vahidR thanks!
